import-service/setup_gae_firewall.py

```diff
@@ -0,0 +1,148 @@
+###
+print(f"\nSetting Broad firewall rules on project {args.project}.")
+
+print("Setting default firewall rule to DENY")
+cmd = f"gcloud --project {args.project} app firewall-rules update default --action deny".split()
+res = subprocess.run(cmd, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, encoding='utf-8')
+check_error(res, "Error updating default firewall rule, exiting")
+
+for (idx, ip_range) in enumerate(broad_range_cidrs):
+    add_gae_firewall_rule(1000 + idx, ip_range, "Broad internal network")
+
+ip_count = len(broad_range_cidrs)
+
+###
+# Add rule to let Pub/Sub traffic through.
+###
+print(f"\nSetting Pub/Sub firewall rule.")
+add_gae_firewall_rule(1000 + ip_count, PUBSUB_IP_RANGE, "GCP Pub/Sub")
+
+ip_count += 1
+
+
+def get_gae_public_ips(gcloud_instances_list):
+    """Parses out the monstrous gcloud output into a dict of instance name -> IP.
+    For reference, gcloud returns a list of these: https://cloud.google.com/compute/docs/reference/rest/v1/instances"""
+    all_instance_nics = {inst["name"]:inst["networkInterfaces"] for inst in gcloud_instances_list}
+    return {inst:ac["natIP"]
+            for inst in all_instance_nics
+            for nic in all_instance_nics[inst]
+            for ac in nic["accessConfigs"]
+            if nic["kind"]=="compute#networkInterface" and ac["kind"]=="compute#accessConfig"}
+
+
+###
+# Add back-Rawls rules.
+###
+print(f"\nAdding firewall rule for back-rawls.")
+
+# List all the back-rawls instances so we can get their IPs. There should be only one.
+cmd = f'gcloud --project broad-dsde-{args.env} compute instances list'.split() + [f'--filter=name:gce-rawls-{args.env}* AND status:RUNNING AND tags.items=backend', '--format=json']
+res = subprocess.run(cmd, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, encoding='utf-8')
+check_error(res, "Error finding back-rawls, exiting")
+
+back_rawlses = json.loads(res.stdout)
+if len(back_rawlses) != 1:
+    print(f"Found {len(back_rawlses)} back-rawlses, expecting 1")
+    sys.exit(1)
+
+rawls_inst_ips = get_gae_public_ips(back_rawlses)
+for (idx, rawls_inst) in enumerate(rawls_inst_ips):
+    add_gae_firewall_rule(1000 + ip_count + idx, rawls_inst_ips[rawls_inst], rawls_inst)
+
+ip_count += len(rawls_inst_ips)
+
+
+###
+# Add Orchestration rules.
+###
+print(f"\nAdding firewall rules for Orch instances.")
+
+# List all the orch instances. There will be many.
+cmd = f'gcloud --project broad-dsde-{args.env} compute instances list'.split() + [f'--filter=name:gce-firecloud-orchestration-{args.env}* AND status:RUNNING', '--format=json']
+res = subprocess.run(cmd, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, encoding='utf-8')
+check_error(res, "Error finding orch instances, exiting")
+
+orchestrations = json.loads(res.stdout)
+
+orch_inst_ips = get_gae_public_ips(orchestrations)
+for (idx, orch_inst) in enumerate(orch_inst_ips):
+    add_gae_firewall_rule(1000 + ip_count + idx, orch_inst_ips[orch_inst], orch_inst)
+
```

Comment:
It would be ideal if this script collected all of the deletes and creates it intends to issue, then only applies those at the end. Else, an error mid-script could result in deleting all firewall rules and re-adding none, leaving things in an ugly state until  everything is debugged and fixed. I'm trying to decide if this is worth holding the PR for ...  what are your thoughts?

Reply:
it's probably not a hugely intrusive change to the script to do so -- just means generating the list of rawls and orch IPs ahead of the first delete. i'll see what i can do

Reply:
done